arangod/Aql/OptimizerRules.cpp

```diff
@@ -7261,43 +7265,317 @@ void arangodb::aql::moveFiltersIntoEnumerateRule(Optimizer* opt, std::unique_ptr
+      return false;
+    // UnsortingGather currently does not work. Also, as we would possibly add
+    // them with remote nodes in the query, we exclude these here, too.
+    case ExecutionNode::REMOTE:
+    case ExecutionNode::GATHER:
+    case ExecutionNode::NORESULTS:
+      // no results currently cannot work, as they do not fetch from above.
+    case ExecutionNode::LIMIT:
+      // limit blocks currently cannot work, both due to skipping and due to the
+      // limit and passthrough, which forbids passing shadow rows.
+      return true;
+    case ExecutionNode::COLLECT: {
+      auto const collectNode = ExecutionNode::castTo<CollectNode const*>(node);
+      // Collect nodes skip iff using the COUNT method.
+      return collectNode->aggregationMethod() == CollectOptions::CollectMethod::COUNT;
+    }
+    // TODO Enable modification nodes again, as soon as the corresponding branch
+    //  is merged.
+    case ExecutionNode::INSERT:
+    case ExecutionNode::REMOVE:
+    case ExecutionNode::REPLACE:
+    case ExecutionNode::UPDATE:
+    case ExecutionNode::UPSERT:
+      // These should already have been handled
+      TRI_ASSERT(false);
+      return true;
+    case ExecutionNode::MAX_NODE_TYPE_VALUE:
+      break;
+  }
+  THROW_ARANGO_EXCEPTION_FORMAT(
+      TRI_ERROR_INTERNAL_AQL,
+      "Unhandled node type '%s' in splice-subqueries optimizer rule. Please "
+      "report this error. Try turning off the splice-subqueries rule to get "
+      "your query working.",
+      node->getTypeString().c_str());
+}
+
+void findSubqueriesSuitableForSplicing(ExecutionPlan const& plan,
+                                       containers::SmallVector<SubqueryNode*>& result) {
+  TRI_ASSERT(result.empty());
+  using ResultVector = decltype(result);
+  using BoolVec = std::vector<bool, short_alloc<bool, 64, alignof(size_t)>>;
+
+  using SuitableNodeSet = std::set<SubqueryNode*, std::less<SubqueryNode*>, short_alloc<SubqueryNode*, 128, alignof(SubqueryNode*)>>;
+
+  // This finder adds subqueries that are suitable for splicing. That means that
+  // neither the containing subquery skips - at least not in an ancestor of the
+  // subquery - nor the subquery skips inside.
+  //
+  // It will be used in a fashion where the recursive walk on subqueries is done
+  // *before* the recursive walk on dependencies.
+  // It maintains a stack of bools for every subquery level. The topmost bool
+  // holds whether we've encountered a skipping block so far.
+  // When leaving a subquery, we decide whether it is suitable for splicing by
+  // inspecting the two topmost bools in the stack - the one belonging to the
+  // insides of the subquery, which we're going to pop right now, and the one
+  // belonging to the containing subquery.
+  //
+  // *All* subquery nodes will be added to &result in pre-order, and all
+  // *suitable* subquery nodes will be added to &suitableNodes. The latter can
```

Comment:
I assume the suitable nodes are in the same order in `suitableNodes` as they are in `result`?

Reply:
The code also seems to remove the unsuitable nodes from result? Then suitablenodes and result are the same? (except for one being a set and the other a vector?)

Reply:
> I assume the suitable nodes are in the same order in suitableNodes as they are in result?
No, that's the point (and `suitableNodes` is a set, so I don't control the order - it's just pointers). At the point where I can add nodes in the right order (each subquery must occur before contained subqueries), I do not know yet whether the node is suitable. Because of that I effectively remember the order of (all) nodes in `result`, and which nodes I actually want in `suitableNodes` later.

> The code also seems to remove the unsuitable nodes from result? Then suitablenodes and result are the same? (except for one being a set and the other a vector?)
Yes. In the end, the result must contain exactly the suitable nodes, but in a specific order.